site/docs/api/error-handling.md

```diff
@@ -0,0 +1,136 @@
+
+So even if we assume no bugs in the Jet framework itself, errors can and
+will happen. What options Jet offers for dealing with them is the topic
+of the next sections.
+
+## Scope
+
+A Jet cluster, at any given time, is running an arbitrary number of
+independent jobs in parallel. These jobs do not interact in any way, and
+the failure of one should not lead to any consequences for any other.
+When considering error handling, the basic unit of discussion will be
+the job, mainly what happens with it once it encounters a problem.
+
+The default way Jet reacts to an error is to fail the job that has
+produced it, but what one can do with such a failed job afterward is the
+interesting part.
+
+## Types of Errors
+
+Before we dig deeper, let's list some typical error one might encounter:
+
+* *corrupt data*: ex. user-provided code is parsing JSON lines, but some
+  of them are incorrectly formatted
+* *bugs in user code*: ex. user provided filtering lambda throws a
+  `NullPointerException` for certain input
+* *connection to external services lost*: Jet sink writing into an
+  `IMap` loses connection to the remote Hazelcast cluster hosting the
+  data
+
+## Specific Solutions
+
+Some concrete types of error have solutions specifically tailored to
+address them.
+
+### In-Memory Data Structures
+
+Jet comes out of the box with some [in-memory distributed data
+structures](data-structures) which can be used as a data source or a
+sink. This are hosted by Hazelcast clusters, which can be the same as
+the one Jet runs on or separate, remote ones.
+
+In this case, all error handling (for example, communications failures
+with the remote cluster, but not just) is handled transparently to Jet,
+by the Hazelcast cluster itself. For further details and configuration
+options, read the [relevant section in the Hazelcast
+manual](https://docs.hazelcast.org/docs/latest/manual/html-single/#handling-failures).
+
+### Individual Sources & Sinks
+
+Sources and Sinks are Jet's points of contact with external data stores.
+Their various types have specific characteristics that enable error
+handling strategies applicable only to them.
+
+For example, our [Change Data Capture
+sources](sources-sinks#change-data-capture-cdc) can attempt to reconnect
+automatically whenever they lose connection to the databases they
+monitor, so for intermittent network failures, their owner jobs don't
+need to fail.
+
+For further such specific failure handling options, consult the Javadoc
+of Jet's various [sources and sinks](sources-sinks).
+
+## Generic Solutions
+
+In Jet we also have generic options for dealing with random errors
+affecting our jobs.
+
+### Jobs Without Mutable State
+
+For many streaming jobs, specifically the ones which don't have any
+processing guarantee configured, the pipeline definition and the job
+config are the only parts we can identify as state, and those are
+immutable.
+
+Batch jobs also fall into the category of immutable state jobs.
```

Comment:
You are perfectly right, theoretically. Batch jobs do have state but they don't actually save it, so in practical terms they don't have state... You shouldn't even enable snapshotting for batch jobs, because there are known scenarios when it wouldn't work well (for example batch job with keyed windowing). Viliam can list a ton of related issues just from the top of his head... I will try to find the right formulation.